src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -259,34 +264,23 @@ internal JsonTypeInfo(Type type, JsonConverter converter, Type runtimeType, Json
                                     // Non-public fields should not be included for (de)serialization.
                                 }
                             }
-                        }
 
-                        JsonPropertyInfo[] cacheArray;
-                        if (DetermineExtensionDataProperty(cache))
-                        {
-                            // Remove from cache since it is handled independently.
-                            cache.Remove(DataExtensionProperty!.NameAsString);
+                            currentType = currentType.BaseType;
+                            if (currentType == null)
+                            {
+                                break;
+                            }
 
-                            cacheArray = new JsonPropertyInfo[cache.Count + 1];
+                            properties = currentType.GetProperties(bindingFlags);
+                        };
 
-                            // Set the last element to the extension property.
-                            cacheArray[cache.Count] = DataExtensionProperty;
-                        }
-                        else
+                        if (DetermineExtensionDataProperty())
```

Comment:
Instead of checking at the end (and looping through all the properties again), should we check for the attribute and validate uniqueness during the first pass to create `JsonPropertyInfo`s?

Reply:
Yes it could use a little love here so I did refactor this (although it's been like this for a long time).